plutus-core/executables/pir/Main.hs

```diff
@@ -1,106 +1,288 @@
+  <> short 'l'
+  <> metavar "LANGUAGE"
+  <> value UPLC
+  <> showDefault
+  <> help ("Target language: plc or uplc")
+  )
+
+pOptimise :: Parser Bool
+pOptimise = switch'
+            (  long "dont-optimise"
+            <> long "dont-optimize"
+            <> help ("Turn off optimisations")
+            )
+
+pJustTest :: Parser Bool
+pJustTest = switch ( long "test"
+                   <> help "Just report success or failure, don't produce an output file"
+                   )
+
+pCompileOptions :: Parser CompileOptions
+pCompileOptions = CompileOptions
+               <$> pLanguage
+               <*> pOptimise
+               <*> pJustTest
+               <*> input
+               <*> pPirInputFormat
+               <*> output
+               <*> outputformat
+               <*> printmode
+
+pPirOptions :: Parser Command
+pPirOptions = hsubparser $
+             command "analyse"
+                  (info (Analyse <$> pAnalyseOptions) $
+                   progDesc $
+                   "Given a PIR program in flat format, deserialise and analyse the program, " <>
+                   "looking for variables with the largest retained size.")
+           <> command "compile"
+                  (info (Compile <$> pCompileOptions) $
+                   progDesc $
+                   "Given a PIR program in flat format, deserialise it, " <>
+                   "and test if it can be successfully compiled to PLC.")
+           <> command "convert"
+                  (info (Convert <$> convertOpts)
+                   (progDesc $
+                    "Convert a program between various formats" <>
+                    "(only 'textual' (default) and 'flat-named' are available for PIR)."))
+           <> command "optimise"
+                  (info (Optimise <$> pPirOptimiseOptions)
+                   (progDesc "Run the PIR optimisation pipeline on the input."))
+           <> command "optimize"
+                  (info (Optimise <$> pPirOptimiseOptions)
+                   (progDesc "Run the PIR optimisation pipeline on the input."))
+           <> command "print"
+                  (info (Print <$> printOpts) $
+                 progDesc $
+                   "Given a PIR program in textual format, " <>
+                   "read it in and print it in the selected format.")
+
+
+-- | Load a PIR program (in either textual of flat-named format)
+getPirProgram ::
+    PirFormat ->
+    Input ->
+    IO (PirProg PLC.SrcSpan)
+getPirProgram fmt inp =
+    case fmt of
+        TextualPir -> snd <$> parseInput inp
+        FlatNamed -> do
+            prog <- loadTplcASTfromFlat Named inp  :: IO (PirProg ())
+            -- No source locations in Flat, so we have to make them up.
+            return $ topSrcSpan <$ prog
+
+-- | Write a PIR program, but only with the restricted types supported by PIR
+writePirProgram ::
+    Output ->
+    PirFormat ->
+    PrintMode ->
+    PirProg () ->
+    IO ()
+writePirProgram outp pirFmt mode prog =
```

Comment:
maybe this is similar to [writeProgram](https://github.com/input-output-hk/plutus/blob/3bffea1f8738077874dcb859997b9f2970a9be9a/plutus-core/executables/src/PlutusCore/Executable/Common.hs#L469)?

Reply:
Yes,but there are problems with the other formats (de Bruijn etc) when you do that: the help says you can use them then you get a runtime error if you try it.  I'm going to try adding the other formats for PIR too so that we can use the same parsers for everything (which is why I attached the big notice saying "**PLEASE DON'T REVIEW THIS YET"** at the top of the PR!)

Reply:
Sorry! I thought that was outdated, it told me you asked for a review.

Reply:
> Sorry! I thought that was outdated, it told me you asked for a review.


I did ask for reviews but then removed the reviewers when I decided to try to fix the input format thing, but it seems that GitHub doesn't say anything to the reviewers when you do that.  Next time maybe I should warn the reviewers on Slack.

Anyway, thanks for the comments! I might ask for more later...